### PR TITLE
Include the Sim Error Activity Type Stack trace

### LINKED ISSUE #1415: Include activity names + activity hierarchy with sim error messages
### Checked for duplicates

Yes - I've already checked

### Alternatives considered

Yes - and alternatives don't suffice

### Related problems

This builds on the work done in #659 to improve error messages when the simulation encounters an error.

### Describe the feature request

After reviewing with @parkerabercrombie - the improvements in #659 are helpful for tracking down activities which cause an error, but we identified two more desired features for improving the user debugging experience:

* The error message should include the name of the activity, not just its ID.
* If the error occurs in a child/descendant activity, then the error message should include not just the activity itself but the hierarchy of its parent/ancestors - sort of a "stack trace" of activities.

### PR DESCRIPTION
**Pull request by goetzrrGit**
* **Tickets addressed:** Closes #1415 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
This PR provides more detailed error information and improves debugging capabilities for Clipper. The error message includes the name of the affected activity type and provides a stack trace-like representation of the activity type hierarchy when errors occur in child activities.

```json
{
  "utcTimeDoy": "2023-001T01:02:00",
  "elapsedTime": "01:02:00.000000",
  "activityStackTrace": 

        "DaemonTaskActivity
        |-DaemonCheckerSpawner
        |--DaemonCheckerActivity",

  "executingDirectiveId": 5,
  "executingActivityType": "DaemonTaskActivity"
}

```

## Verification
e2e test pass and I am including a new parent activity, which will be used to spawn children 2 levels deep. This allows for the error message to have a more detailed stack trace.

### CODE REVIEW

File: merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/SimulationDriver.java [hunk 1]
```diff
@@ -110,8 +110,9 @@ public static <Model> SimulationResults simulate(
         // Swallowing the spanException as the internal `spanId` is not user meaningful info.
         final var topics = missionModel.getTopics();
         final var directiveId = engine.getDirectiveIdFromSpan(activityTopic, topics, ex.spanId);
+        final var activityTrace = engine.findActivityTraceFromSpan(engine, activityTopic, topics, ex.spanId);
```
Comment (Mythicaeda) [inline merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/SimulationDriver.java]: Minor, repeated for `CheckpointSimDriver` and `CachedSimEngine`: It looks like the trace should only calculated if `directiveId.isPresent()`, so it ought to go inside the `if` block.

File: e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SimulationTests.java [hunk 2]
```diff
@@ -470,7 +470,9 @@ void directiveIdIncluded() throws IOException {
 
       // The error includes the directive id that was executing
       assertTrue(exception.containsKey("executingDirectiveId"));
+      assertTrue(exception.containsKey("executingActivityType"));
       assertEquals(dirId, exception.getInt("executingDirectiveId"));
+      assertEquals("DaemonCheckerActivity", exception.getString("executingActivityType"));
```
Comment (Mythicaeda) [inline e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SimulationTests.java:476]: `activityStackTrace` needs to be checked here

File: merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java [hunk 3]
```diff
@@ -676,6 +677,45 @@ void extractOutput(final SerializableTopic<T> topic, final Event ev, final SpanI
...
+   *
+   * This method analyzes the provided span within the event graph to identify a list of parent activities leading up to it.
+   *
+   * @param engine the simulation engine
+   * @param activityTopic the activity topic
+   * @param serializableTopics a collection of serializable topics
+   * @param spanId the ID of the span to analyze
+   * @return a list of serialized activities representing the stack trace
+   */
+  public List<SerializedActivity> findActivityTraceFromSpan(
```
Comment (Mythicaeda) [inline merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java]: The code for this and `getDirectiveIdFromSpan` overlap quite a bit. Is it possible to collapse them?

File: merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/SimulationDriver.java (hunk 1 above)
Comment (goetzrrGit) [reply merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/SimulationDriver.java]: I reworked things and merged the `getDirectiveIdFromSpan` and `findActivityTraceFromSpan` into one function call `getDirectiveDetailFromSpan`. This function returns a newly created [`DirectiveDetail`](https://github.com/NASA-AMMOS/aerie/pull/1518/files#diff-e56006b03746946a05e6908ae32737b0e4c96d1a63e86b885f5512bcc87103c0)

File: e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SimulationTests.java (hunk 2 above)
Comment (goetzrrGit) [reply e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SimulationTests.java:476]: Updated the e2e test to include the missing `actiivtyStackTrace` from this test.

File: merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java (hunk 3 above)
Comment (goetzrrGit) [reply merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java]: I reworked things and merged the `getDirectiveIdFromSpan` and `findActivityTraceFromSpan` into one function call `getDirectiveDetailFromSpan`. This function returns a newly created [`DirectiveDetail`](https://github.com/NASA-AMMOS/aerie/pull/1518/files#diff-e56006b03746946a05e6908ae32737b0e4c96d1a63e86b885f5512bcc87103c0)

File: merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java [hunk 4]
```diff
@@ -676,29 +677,36 @@ void extractOutput(final SerializableTopic<T> topic, final Event ev, final SpanI
...
     Optional<SpanId> directiveSpanId = Optional.of(spanId);
+    final var activityStackTrace = new LinkedList<SerializedActivity>();
     while (directiveSpanId.isPresent() && !spanInfo.isDirective(directiveSpanId.get())) {
+      activityStackTrace.add(spanInfo.input().get(directiveSpanId.get()));
       directiveSpanId = this.getSpan(directiveSpanId.get()).parent();
     }
-    return directiveSpanId.map(spanInfo::getDirective);
+
+    // Add final top level parent activity to the stack trace if present
+    if( directiveSpanId.isPresent()){
```
Comment (Mythicaeda) [inline merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java]: Nitpick, nonblocking: spacing
```suggestion
    if (directiveSpanId.isPresent()) {
```

**Review by Mythicaeda — APPROVED** [review]

**Pull request merged**